rfcs/20210307-tfx-api-artifact-compat.md

```diff
@@ -0,0 +1,583 @@
+implementations of an official TFX component, we plan to deprecate this support
+externally and declare it internal only. We will only support currently known
+use cases listed above.
+
+#### Scaffolding Templates {#scaffolding-templates}
+
+The scaffolding template CLI command itself is still **experimental**, therefore
+excluded from compatibility compliance. Note that generated pipeline code &
+config are considered user code, therefore they can expect the same
+compatibility compliance as listed above.
+
+Putting it differently:
+
+*   The structure of code and configs generated is subject to (possibly
+    breaking) change between minor versions;
+*   Once generated, the produced pipeline code and configs are essentially user
+    code which relies on TFX’s public API, thus guaranteed to work with newer
+    minor versions within the same major version, as explained in backward
+    compatibility rules.
+    *   This rule ensures that even if a user upgraded the `tfx` package in
+        their environment, the scaffolded pipeline can still function.
+
+#### Examples, Notebooks and Tutorials {#examples-notebooks-and-tutorials}
+
+Examples, Notebooks (Colab) and Tutorials are **excluded** from the
+compatibility compliance. These are created mostly for demonstrating how to use
+TFX in accordance with best practices, therefore often subject to a particular
+version of TFX.
+
+These examples should use the stable public API of `tfx` as much as possible.
+Usage of experimental or deprecated APIs in these examples should be explicitly
+called out, and the examples should pin their dependency version of `tfx` down
+to a narrow range as much as possible in these rare cases.
+
+#### New Orchestrator {#new-orchestrator}
+
+The new orchestrator is in early stages of development and we don’t think it’ll
+be production-ready for 1.0 timeline, thus it will enter 1.0 as `experimental`.
+
+#### Portability Layer to Other Runners (aka Orchestrators) {#portability-layer-to-other-runner-orchestrators}
+
+This portability layer in tfx refers to various DagRunners which runs TFX
+pipelines on various external orchestrators, the _portable_ module and the
+abstraction between CLI and Dag runner. These libraries will comply with
+backward compatibility.
+
+Notice that several Runners are experimental by their nature, or slated to be
+deprecated. Specifically, we plan to support the following runners with backward
+compatibility:
+
+*   LocalDagRunner
+*   KubeflowV2DagRunner
+
+The following Dag runners will remain experimental:
+
+*   AirflowDagRunner
+    *   Honestly speaking, support of this one is on the fence. Due to lack of
+        proper configuration for a distributed backend support, it’s not much
+        more powerful than the LocalDagRunner when used out-of-box. However,
+        some partner teams managed to connect this to a distributed backend.
+        Until that is supported in TFX, we don’t think this one is stable.
```

Comment:
> Distributed backend

Airflow does support quite well distributing workloads via the KubernetesExecutor, Celery or even KEDA. Not sure what this is referring to. 

Note that Airflow usage is quite large at this point in the idustry with many managed options to make it distributed (including Cloud Composer). 

Also it does allow for better customization than the Kubeflow integration as of today (retries, failure callbacks...)

Reply:
> Not sure what this is referring to.

This is mostly referring our current implementation, which only uses default PythonOperator to run nodes locally.

If there enough community interest, we are happy to partner with the OSS community to build up the Airflow orchestration and utilize all the benefits you mentioned above, but I don't think we have enough resources and expertise in TFX team to do this ourselves.

Reply:
I mean, it works already, you just need to configure Airflow to use KubernetesExecutor. No need to change any part of the integration (which is why it confuses me)

Reply:
Thanks for pointing that out. We are not really aware of this.

Reply:
Would love to sync on this and explain a bit.

Reply:
Proposed rewording:

```
Due to low utilization of the DagRunner, we want to avoid calling it stable. Support will continue to graduate it into production as utilization evolves.
```

Mostly want to avoid calling out distributed backend support. Cloud Composer does support distributed backend  and such, so would avoid saying `Due to lack of proper configuration for a distributed backend support`. I'm okay with the reason being: `Due to low usage and it not being a production usage yet that we dont believe its stable`.

Reply:
Accepted.